### PR TITLE
[https://nvbugs/6676312][fix] Restore DSA CUDA TopK workspace lifecycle

### PR DESCRIPTION
**Pull request by lfr-0531**
## Dev Engineer Review

- DSA metadata owns persistent heuristic and Radix TopK workspaces.
- Workspace capacity covers linear, static-tree, and dynamic-tree speculative decoding.
- `TopK.forward` accepts caller-owned buffers through `radix_ext_kwargs`.
- TopK validates and slices buffers for supported Radix paths.
- Hidden process-global workspace allocation was removed.
- No configuration or test-list files changed.
- No dependency changes were introduced.
- CI requires follow-up because the main pipelines and some helper jobs failed.

## QA Engineer Review

- Added `test_use_self_sampling_gvr`.
- Added `test_use_gvr_emission_threads_to_params`.
- Added `test_use_gvr_emission_defaults_off`.
- Added `test_indexer_two_level_gvr_dispatch`.
- Added `test_metadata_spec_update_resizes_radix_workspace`.
- Updated metadata cache geometry coverage to include `use_self_sampling`.
- Updated TopK tests for caller-owned workspaces and CUDA Graph capture and replay isolation.
- The modified tests were not identified in `tests/integration/test_lists/`, `test-db/`, or `qa/`.
- The tests provide focused regression coverage, but test-list coverage needs follow-up.
- Verdict: needs follow-up.

## Description

Commit `4f13faf713` moved DeepSeek-V4 DSA native CUDA TopK scratch from the
metadata-owned eager/CUDA-graph lifecycle into TopK's generic process-global
arena. A captured graph could therefore retain workspace storage whose
ownership and capacity were not tied to that graph metadata.

This PR restores caller-owned Radix scratch without introducing a workspace
class. DSA metadata creates persistent `radix_aux_indices` and
`radix_aux_logits` tensors with its existing eager/CUDA-graph buffer lifecycle
and resizes them for the maximum reachable speculative row count. Indexer
passes them through `radix_ext_kwargs`; TopK validates and slices them for the
direct CUDA Radix path and native Radix fallbacks from the CuTe DSL paths.

After #18446 landed, the branch was rebased and its superseded CUDA GVR and
heuristic-scratch changes were removed. The new two-level CuTe DSL GVR dispatch
from `main` is retained unchanged. Workspace sizing covers linear, static-tree,
and dynamic-tree speculative geometries. Rebased onto `main` at `d7ae24756e`.

There are no user-facing API or dependency changes.

## Test Coverage

- Exact GB300 culprit/parent reproduction with identical model, native runtime,
  topology, and Agentperf workload:
  - culprit `4f13faf713`: 131/1140 BOS-spam responses (11.49%), average OSL
    690.81, 26.14% truncated.
  - parent `0e00a9481e`: 6/1611 (0.37%), average OSL 316.67, 3.23% truncated.
- Exact culprit plus the metadata-owned implementation, repeated on independent
  GB300 node pairs with new server processes:
  - job `653668`: 1/161 BOS-spam (0.62%), average OSL 360.91, 6.21% truncated.
  - job `653669`: 0/125 BOS-spam, 0 BOS markers, average OSL 301.86,
    4.00% truncated.
  - combined: 1/286 BOS-spam responses (0.35%).
- Post-#18446 rebase GB300 focused validation with the matching SM103 native
  build (job `677418`):
  - DSA metadata/workspace selection: 4 passed.
  - full `test_top_k.py`: 17 passed, including CUDA Graph replay with stable
    caller-owned Radix workspace.
- MXFP4 model validation on two-node TP8/EP8 GB300 aggregate servers with CUDA
  Graphs and the native Radix path enabled:
  - `deepseek-ai/DeepSeek-V4-Pro-DSpark@7c09739`: 32/32 controlled requests
    succeeded, 32/32 arithmetic answers correct, 0 BOS markers.
  - `deepseek-ai/DeepSeek-V4-Pro-0813@72e1d32`: 32/32 controlled requests
    succeeded, 32/32 arithmetic answers correct, 0 BOS markers.
  - The long-context Agentperf coding dataset exceeded the validation engine's
    4096-token limit and was excluded; it did not reach model execution.
- Changed-file ruff check/format, Python 3.12 source compilation,
  `git diff --check`, semantic `range-diff`, conflict-marker scan, DCO, and
  GitHub pre-commit checks passed.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by longcheng-nv** [comment]
Heads-up on a directional overlap with #18446 (two-level GVR decode top-K dispatch, which also folds in the CUDA-heuristic removal).

#18446's commit 2 removes the CUDA GVR path entirely: it deletes `heuristicTopKDecode.{cu,h}` + `heuristic_topk.cuh`, drops the `CUDA_GVR` enum/branch/workspace from `top_k.py`, and removes `pre_idx`/`heuristic_scratch` from the `indexer_topk_decode` op schema (including `register_fake`). That overlaps directly with this PR — same files (`top_k.py`, `dsa/metadata.py`, `dsa/indexer.py`, `test_top_k.py`, `test_dsa_indexer.py`) and a hard conflict on the op schema, since this PR *adds* caller-owned `heuristic_scratch` while #18446 *removes* it.

This is a valid fix for current main, where `CUDA_GVR` is genuinely reachable. But I checked that under #18446's two-level dispatch it is never selected. On main the indexer picks `CUDA_GVR` for `enable_heuristic_topk=True` when the DSL fast path does not apply (`use_cute_dsl_topk == False`); #18446 reroutes exactly that case to exact radix (`CUTE_DSL_RADIX`/`CUDA_RADIX`) with a one-time `gvr_prereq_radix_fallback` warning, so `CUDA_GVR` drops out of the selector entirely (0 references remain in `tensorrt_llm/`). In effect nvbug 6676312 is resolved by elimination — the buggy workspace path is simply no longer taken.

So if #18446 lands as planned, this PR would be superseded. Could we align on direction before either merges, to avoid rework on the shared files/op schema? If there is a config that still needs the CUDA GVR path, that would contradict #18446's removal premise and we should reconcile that first. cc @longcheng-nv

**Comment by longcheng-nv** [comment]
Correction / narrowing of my earlier comment: #18446 supersedes the **CUDA_GVR** half of this PR (it deletes that path), but **not** the CUDA Radix workspace hardening.

Looking more closely at #18446's `top_k.py`: it keeps the process-global `_memory_buffers` arena for the CUDA Radix split-work scratch (`_get_radix_workspace` -> `get_buffer(..., reserve_buffer=capture_graph)`). That is capture-time reservation — which this PR's own negative ablation flagged as insufficient: you found construction-time maximum-capacity reservation is required, and lazy/at-capture allocation stayed bad (24/91, 25/110 BOS-spam). #18446 also reroutes the former `enable_heuristic_topk` CUDA_GVR configs to CUDA_RADIX, so if the same workspace-lifecycle issue reaches the Radix path, #18446 would not eliminate it — it would put more traffic onto it.

So the accurate picture: #18446 removes the CUDA_GVR manifestation, but the CUDA Radix / native-radix workspace lifecycle fix here is complementary and should be retained (or ported onto the post-#18446 `top_k.py`), not dropped. Which parts overlap vs. remain-needed depends on whether nvbug 6676312's repro was CUDA_GVR-only or also exercised CUDA_RADIX — I'm checking the bug report for that and will follow up.

cc @longcheng-nv @lfr-0531

**Comment by longcheng-nv** [comment]
Stepping back on how this sits next to #18446 (which deletes the CUDA GVR path):

The CUDA_GVR half here is largely overtaken by #18446. On SM100 with DSL available it routes `enable_heuristic_topk` to CUTE_DSL_GVR — no process-global scratch, and the barrier race is already covered by #18339 — so the heuristic path this bug reproduces on stops going through `_memory_buffers` once #18446 lands, and the GVR-specific scratch plumbing retires with it.

The radix scratch is a different story, and I don't think it should retire with the GVR bits. `top_k_radix_indices/values_workspace` is still the live fallback: `_forward_decode_radix` drops from the DSL radix op to the C++ radix tier whenever `compress_ratio > 1 and next_n > 1` (i.e. MTP + compressed decode), and it's the default whenever DSL is unavailable. Same process-global pool, same op-type+device key with no layer/graph/stream identity, same aliasing under graph capture. So the metadata-owned ownership fix for the radix tier stands on its own, independent of the CUDA_GVR decision.

My suggestion: keep the radix workspace ownership fix; let the CUDA_GVR-specific parts land or retire with #18446. If you'd rather get this in ahead of #18446 to close the P0, no harm — those parts just become dead code once the path is removed.

**Comment by lfr-0531** [comment]
Discussed with @longcheng-nv, let's wait for https://github.com/NVIDIA/TensorRT-LLM/pull/18446 to be merged first. Once that's done, I can rebase this PR and keep only the radix workspace ownership fix.

**Review by longcheng-nv — APPROVED** [review]
LGTM. The fix correctly restores the radix workspace ownership to the DSA metadata lifecycle, and the CI/validation results look good. Non-blocking: consider adding a regression test using the actual metadata/buffer lifecycle. Approving.

File: tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py [hunk 1]
```diff
@@ -1105,6 +1133,8 @@ def update_spec_dec_param(
         init_shape = self.kv_lens_expanded_host.shape[0]
         if self.max_num_sequences * (1 + self.max_draft_tokens) != init_shape:
             self.create_expanded_buffers(capture_graph=capture_graph)
+        if resize_radix_workspace:
+            self._create_radix_aux_buffers(capture_graph)
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py:1137]: Can we ensure that this resize only happen once for spec-dec and never happen again?

File: tensorrt_llm/_torch/modules/top_k.py [hunk 2]
```diff
@@ -110,6 +109,8 @@ def forward(
                 self-sampling engine does not consume this state.
                 ``gvr_row_order`` is an optional int32 request ordering with
                 shape ``[num_requests]`` on the same device.
+            radix_ext_kwargs: Caller-owned ``radix_aux_indices`` and
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/top_k.py]: If radix_ext_kwargs always only contains radix_aux_indices and radix_aux_logits, should we unpack it as flatten arguments instead?

File: tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/attention/backends/sparse/dsa/metadata.py:1137]: Yes. `update_spec_dec_param()` runs on every forward, but the workspace size is derived from engine-level maxima, not the runtime batch:
```
        rows_per_sequence = max(1 + max_draft_len, 1 + (max_total_draft_tokens or 0))
        if is_spec_dec_dynamic_tree and spec_tree_manager is not None:
            rows_per_sequence = max(rows_per_sequence, spec_tree_manager._internal_buf_dim)
```
The equality guard therefore allows at most one resize per metadata instance.

File: tensorrt_llm/_torch/modules/top_k.py (hunk 2 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/top_k.py]: Agreed. Unlike the GVR options, the Radix workspace is a fixed tensor pair. I replaced `radix_ext_kwargs` with explicit keyword-only `radix_aux_indices` and `radix_aux_logits` arguments.

**Review by yuxianq — APPROVED** [review]
